Bluetooth: Add configuration to control the dynamic connection callbacks

Yagoor opened the pull request:
* In most use cases the connection callbacks do not change during the lifetime of the application meaning that they can be registered at compile time instead of runtime. This has a smaller code footprint
* Remove references in the stack to bt_conn_cb_{register,unregister}

cvinayak reviewed (changes requested):
We have extensive (my belief) Host, Mesh and LE Audio test coverage in CI, please pick a `tests/bsim/bluetooth/`  test to add a variant to cover this new feature to be tested in PR CI for regressions.

Thalley commented on subsys/bluetooth/audio/bap_unicast_client.c:
```diff
@@ -4467,14 +4467,13 @@ static void unicast_client_disconnected(struct bt_conn *conn, uint8_t reason)
 }
 
-static struct bt_conn_cb conn_cbs = {
+BT_CONN_CB_DEFINE(conn_cbs) = {
```
Could you separate the new Kconfig option and the changes to the calls of `bt_conn_cb_register` into separate commits? Also splitting host from audio?

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -118,7 +118,9 @@ sys_slist_t bt_auth_info_cbs = SYS_SLIST_STATIC_INIT(&bt_auth_info_cbs);
 #endif /* CONFIG_BT_SMP || CONFIG_BT_CLASSIC */
 
 
+#if defined(CONFIG_BT_CONN_DYNAMIC_CALLBACKS)
 static sys_slist_t conn_cbs = SYS_SLIST_STATIC_INIT(&conn_cbs);
+#endif /* CONFIG_BT_CONN_DYNAMIC_CALLBACKS */
```
Correct me if I'm wrong, but shouldn't it be possible to define `conn_cbs` without the guard, and then guard all calls to it with `IS_ENABLED(CONFIG_BT_CONN_DYNAMIC_CALLBACKS)`, and then have the linker remove `conn_cbs` from memory?

Yagoor committed:
Bluetooth: Audio: Conn callbacks registration with BT_CONN_CB_DEFINE

This commit registers the connection callbacks using BT_CONN_CB_DEFINE,
which ensures they are set up at link time rather than runtime.

Signed-off-by: Yago Fontoura do Rosario <yafo@demant.com>

Yagoor committed:
Bluetooth: Host: Conn callbacks registration with BT_CONN_CB_DEFINE

This commit registers the connection callbacks using BT_CONN_CB_DEFINE,
which ensures they are set up at link time rather than runtime.

Signed-off-by: Yago Fontoura do Rosario <yafo@demant.com>

Yagoor committed:
Bluetooth: Add configuration to control the dynamic connection callbacks

This commit allows to the user to control the use or not of dynamic
connection callbacks.
The stack is not using the dynamic connection callbacks so they are
only needed if the applications want to use it.
Setting it to true for backwards compatibility.

With dynamic connection callbacks

           FLASH:      176600 B       512 KB     33.68%
             RAM:       27428 B        64 KB     41.85%
        IDT_LIST:          0 GB        32 KB      0.00%

Without dynamic connection callbacks

           FLASH:      176376 B       512 KB     33.64%
             RAM:       27428 B        64 KB     41.85%
        IDT_LIST:          0 GB        32 KB      0.00%

Signed-off-by: Yago Fontoura do Rosario <yafo@demant.com>

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1967,13 +1977,15 @@ void notify_le_param_updated(struct bt_conn *conn)
-						   conn->le.timeout);
+		SYS_SLIST_FOR_EACH_CONTAINER(&conn_cbs, callback, _node) {
+			if (callback->le_param_updated) {
+				callback->le_param_updated(conn, conn->le.interval,
+							conn->le.latency,
+							conn->le.timeout);
```
Indentation is a bit off here

Yagoor replied:
Done

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2655,6 +2677,14 @@ bt_security_t bt_conn_get_security(const struct bt_conn *conn)
 
 int bt_conn_cb_register(struct bt_conn_cb *cb)
```
For the `bt_conn_cb_register` and `bt_conn_cb_unregister` I actually prefer the `#if defined` because then uses of that when `CONFIG_BT_CONN_DYNAMIC_CALLBACKS=n` will result in a compiletime error, rather than runtime, which I think is preferable to developers. Not sure if we have any specific requirements/guidelines on this though

cvinayak commented:
> The `bt_conn_cb_{register,unregister}` references in the stack has been removed so the entire stack is only using the iterable sections so that logic is tested and some of the current applications are using the bt_conn_cb_register to register application level callbacks so that applications will test the run time logic.

Yes, samples tested in CI are using the run time logic, but I am asking for samples to be tested in CI using the new static logic when using `CONFIG_BT_CONN_DYNAMIC_CALLBACKS=n`. You have `CONFIG_BT_CONN_DYNAMIC_CALLBACKS=y` as default in the KConfig, none of the existing samples under CI is covering your PR changes in CI simulation.

cvinayak commented on subsys/bluetooth/Kconfig:
```diff
@@ -109,6 +109,16 @@ config BT_MAX_CONN
 	  Maximum number of simultaneous Bluetooth connections
 	  supported.
 
+config BT_CONN_DYNAMIC_CALLBACKS
+	bool "Dynamic connection callbacks"
+	depends on BT_CONN
+	default y
+	help
+	  Enable support for dynamic connection callbacks. This allows
+	  applications to register and unregister callbacks at runtime.
+	  If disabled, the application must register all callbacks at
+	  compile time using the BT_CONN_CB_DEFINE() macro.
+
```
Needs release notes entry of this new public feature?

Thalley reviewed (dismissed):
I'm OK with the changes. 

Did you consider updating and setting `BT_CONN_DYNAMIC_CALLBACKS=n` in the `prj_minimal.conf` for the central_hr and peripheral_hr samples?

cvinayak commented:
> So you want me to take any test there, make a copy of it and in the prj.conf disable the CONFIG_BT_CONN_DYNAMIC_CALLBACKS?

Take a look at https://github.com/zephyrproject-rtos/zephyr/blob/a1f66f05d292ff51c3c6fdc081867bc0a5b232bd/tests/bsim/bluetooth/samples/compile.sh

Add a new `overlay-static_callbacks.conf`  to the `samples/bluetooth/peripheral_hr` sample and cover it in the above compile and have a new script in `tests_scripts` folder.

Yagoor commented:
> > So you want me to take any test there, make a copy of it and in the prj.conf disable the CONFIG_BT_CONN_DYNAMIC_CALLBACKS?
> 
> Take a look at https://github.com/zephyrproject-rtos/zephyr/blob/a1f66f05d292ff51c3c6fdc081867bc0a5b232bd/tests/bsim/bluetooth/samples/compile.sh
> 
> Add a new `overlay-static_callbacks.conf` to the `samples/bluetooth/peripheral_hr` sample and cover it in the above compile and have a new script in `tests_scripts` folder.
> 
> Rationale, I can point users/customers to this sample (and overlay file) to guide them to when they want lower code size application.

Done 😄

cvinayak commented on tests/bsim/bluetooth/samples/compile.sh:
```
tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr_static_callbacks.sh PASSED (0.168 s)
```
Excellent!

cvinayak commented:
@Yagoor A small request/nitpick (commit message change does not make github to re-request approvals), can you add the before and after code/RAM size in the commit message (not a very strong request tho. as I have put the details below anyway in this PR).

Example:
With `cmake -GNinja -DBOARD=nrf52dk/nrf52832 ../../samples/bluetooth/peripheral_hr`

```
[344/344] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      176600 B       512 KB     33.68%
             RAM:       27428 B        64 KB     41.85%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/vich/workspace/zephyrproject/zephyr/build/peripheral_hr/zephyr/zephyr.elf for board: nrf52dk
```

With `cmake -GNinja -DBOARD=nrf52dk/nrf52832 -DEXTRA_CONF_FILE=overlay-static_callbacks.conf ../../samples/bluetooth/peripheral_hr`

```
[342/342] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      176376 B       512 KB     33.64%
             RAM:       27428 B        64 KB     41.85%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/vich/workspace/zephyrproject/zephyr/build/peripheral_hr/zephyr/zephyr.elf for board: nrf52dk
```

Thalley commented on samples/bluetooth/central_hr/overlay-static_callbacks.conf:
Since this is used for testing, I'd actually prefer to put this into `zephyr/tests/bsim/bluetooth/samples/central_hr_peripheral_hr` instead.

Yagoor replied:
@cvinayak what do you think of this?

Moving it would affect the rationale you mentioned before.

"I can point users/customers to this sample (and overlay file) to guide them to when they want lower code size application."

Thalley replied:
Does disabling `CONFIG_BT_CONN_DYNAMIC_CALLBACKS` actually change memory for this sample?

If it can be disabled and still build, then that means that `bt_conn_cb_register` isn't called, which means that the linker could in theory (not sure if it does it in practice) remove everything related to `CONFIG_BT_CONN_DYNAMIC_CALLBACKS` anyway

Yagoor replied:
> Does disabling `CONFIG_BT_CONN_DYNAMIC_CALLBACKS` actually change memory for this sample?
> 
> If it can be disabled and still build, then that means that `bt_conn_cb_register` isn't called, which means that the linker could in theory (not sure if it does it in practice) remove everything related to `CONFIG_BT_CONN_DYNAMIC_CALLBACKS` anyway

It does change memory for this sample. I can explain why.

The behavior you described is only achievable if LTO is enabled because you need optimization across translations units. Meaning that the linker will see that `bt_conn_cb_register` is not called and remove everything related to it. But in practice, LTO is not always enabled so that define can help with that. Without LTO the optimization is performed in each translation unit and the linker just glues things together

@cvinayak has shown it here:

> 
> Example: With `cmake -GNinja -DBOARD=nrf52dk/nrf52832 ../../samples/bluetooth/peripheral_hr`
> 
> ```
> [344/344] Linking C executable zephyr/zephyr.elf
> Memory region         Used Size  Region Size  %age Used
>            FLASH:      176600 B       512 KB     33.68%
>              RAM:       27428 B        64 KB     41.85%
>         IDT_LIST:          0 GB        32 KB      0.00%
> Generating files from /home/vich/workspace/zephyrproject/zephyr/build/peripheral_hr/zephyr/zephyr.elf for board: nrf52dk
> ```
> 
> With `cmake -GNinja -DBOARD=nrf52dk/nrf52832 -DEXTRA_CONF_FILE=overlay-static_callbacks.conf ../../samples/bluetooth/peripheral_hr`
> 
> ```
> [342/342] Linking C executable zephyr/zephyr.elf
> Memory region         Used Size  Region Size  %age Used
>            FLASH:      176376 B       512 KB     33.64%
>              RAM:       27428 B        64 KB     41.85%
>         IDT_LIST:          0 GB        32 KB      0.00%
> Generating files from /home/vich/workspace/zephyrproject/zephyr/build/peripheral_hr/zephyr/zephyr.elf for board: nrf52dk
> ```

Thalley replied:
Given that RAM is consistent, doesn't that mean that at least `static sys_slist_t conn_cbs` is being removed by the linker in either case? 

I'm also surprised how much FLASH difference there is; shouldn't it only affect the `bt_conn_cb_register` and `bt_conn_cb_unregister` functions?

cvinayak replied:
> There are numerous Kconfig options that a user can change for any sample, and we definitely don't want a .conf file for each of them

Use of dynamic or static callback is a feature of the stack, and the sample demonstrates its use, hence the extra conf file is justified, i.e.  similar to use (or not) of advertising advertising features in this sample.

> Does disabling CONFIG_BT_CONN_DYNAMIC_CALLBACKS actually change memory for this sample?
Yes, ~250 bytes of code space reduced.

>  `bt_conn_cb_register` isn't called, which means that the linker could in theory (not sure if it does it in practice) remove everything related to CONFIG_BT_CONN_DYNAMIC_CALLBACKS anyway

True, but not all of the code related to dynamic callbacks are in direct control path inside `bt_conn_cb_register()` callback, hence linker cannot optimize code that depends on runtime values (say, reading a RAM section for the list of callbacks).

Thinking aloud, I see no RAM reduction, does that mean the callback list is present both in FLASH and RAM (loaded at boot), and RAM section gets a merged list of static and dynamic callback lists? (I am not familiar with Zephyr Linker Script design).

alwa-nordic replied:
> the linker could in theory (not sure if it does it in practice) remove everything related to CONFIG_BT_CONN_DYNAMIC_CALLBACKS anyway

Yes, but I'm pretty sure GCC is not smart enough even with all the LTO flags currently.

Thalley reviewed (dismissed):
Minor request, otherwise LGTM

Yagoor commented on subsys/bluetooth/Kconfig:
```diff
@@ -109,6 +109,16 @@ config BT_MAX_CONN
+	  Enable support for dynamic connection callbacks. This allows
+	  applications to register and unregister callbacks at runtime.
+	  If disabled, the application must register all callbacks at
+	  compile time using the BT_CONN_CB_DEFINE() macro.
```
Done

alwa-nordic commented on subsys/bluetooth/Kconfig:
```diff
@@ -109,6 +109,17 @@ config BT_MAX_CONN
+	depends on BT_CONN
+	default y
+	help
+	  Enable support for dynamic connection callbacks. This links in
+	  bt_conn_cb_register(). Disabling this config is safe. It will make
+	  uses of this and associated APIs fail to link. Note: Static registration
+	  through BT_CONN_CB_DEFINE() is always available and does not depend on
+	  this config.
```
Please, let's keep it as it is. I asked for this description https://github.com/zephyrproject-rtos/zephyr/pull/93703#discussion_r2251368642. I don't think that it reads like a change log is a good argument for shortening the help text. I think it also reads like a help text and will be useful.

This config is default y, so it makes sense to me to include the info that disabling it is safe, since it was not the users choice to enable it in the first place.

alwa-nordic replied:
> I would argue that any Kconfig option is safe to disable, unless otherwise specified :)

I would not trust our Kconfig to be that nice. I suspect a lot of configs will break a system in weird and wonderful ways without any compile-time errors. :smiling_face_with_tear:

jhedberg commented on subsys/bluetooth/Kconfig:
```diff
@@ -109,6 +109,17 @@ config BT_MAX_CONN
 	  Maximum number of simultaneous Bluetooth connections
 	  supported.
 
+config BT_CONN_DYNAMIC_CALLBACKS
+	bool "Dynamic connection callbacks"
+	depends on BT_CONN
```
There's a separate `if BT_CONN` section a little bit further down. I think it makes sense to move this there, and then you can remove the `depends on BT_CONN` since it becomes redundant.

Yagoor replied:
Sounds good, just a question. Is there any difference in using `depends on` vs a `if block`?
